src/ast/module.cc

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "module.h"
+
+#include "path.h"
+
+using namespace peg::udl;
+
+namespace
+{
+  using namespace module;
+
+  ModulePtr make_module(const std::string& name)
+  {
+    return std::make_shared<Module>(name);
+  }
+
+  bool moduledef(ast::Ast& ast, const std::string& path, err::Errors& err)
+  {
+    std::vector<ast::Ast> defs;
+
+    for (auto& node : ast->nodes)
+    {
+      if (node->tag == "moduledef"_)
+        defs.push_back(node);
+    }
+
+    if (defs.size() > 1)
+    {
+      err << defs.front() << "This module contains multiple definitions.\n";
+
+      for (auto& def : defs)
+        err << def << "Module definition here.\n";
+
+      err << err::end;
+      return false;
+    }
+
+    auto moduledef = (defs.size() > 0) ? defs.front() : ast;
+    auto classdef = ast::node(moduledef, "classdef");
+    auto id = ast::token(classdef, "id", "$module");
+    ast::push_back(classdef, id);
+
+    if (moduledef->tag == "moduledef"_)
+    {
+      ast::remove(moduledef);
+
+      for (auto& node : moduledef->nodes)
+        ast::push_back(classdef, node);
+
+      moduledef->nodes.clear();
+    }
+    else
+    {
+      auto typeparams = ast::node(moduledef, "typeparams");
+      ast::push_back(classdef, typeparams);
+      auto oftype = ast::node(moduledef, "oftype");
+      ast::push_back(classdef, oftype);
+      auto constraints = ast::node(moduledef, "constraints");
+      ast::push_back(classdef, constraints);
+    }
+
+    auto typebody = ast::node(moduledef, "typebody");
+    ast::push_back(classdef, typebody);
+
+    for (auto& node : ast->nodes)
+      ast::push_back(typebody, node);
+
+    ast->nodes.clear();
+    ast = classdef;
+    return true;
+  }
+
+  void extract(ast::Ast& ast, std::vector<std::string>& deps)
+  {
+    switch (ast->tag)
+    {
+      case "module_ref"_:
+      {
+        auto name = path::join(ast->path, ast->nodes.front()->token);
+
+        if (path::extension(name).empty())
+          name = path::to_directory(name);
+
+        deps.push_back(name);
+        return;
+      }
+    }
+
+    ast::for_each(ast, extract, deps);
+  }
+
+  ModulePtr load(
+    peg::parser& parser,
+    const std::string& path,
+    const std::string& ext,
+    err::Errors& err)
+  {
+    std::map<std::string, ModulePtr> modules;
+    std::vector<ModulePtr> stack;
+    bool ok = true;
+
+    auto canonical_path = path::canonical(path);
+    auto m = make_module(path);
```

Comment:
There should be a check here for `dep->ast == nullptr`

Reply:
There's no `dep` on line 104. Did you mean line 137? If so, there's no need for `dep->ast` to be valid - `err` will already have the parse error(s), and the module DAG code will work either way.

Reply:
Sorry wrong location. But `sym::build` segfaults due to a null AST when an invalid module path is given.